form.py

```diff
@@ -285,64 +310,319 @@ def open_config(file):
             entries.append(EntryInfo.from_string(line))
     return ConfigInfo(url, entries)
 
-# Returns the passed config file name
-def parse_arguments(argv):
-    import sys
-
-    if len(argv) > 2:
-        print("Too many arguments. Usage: python form.py <filename>")
-        sys.exit(1)
+# Each type has their unique question classes
+QUESTION_CLASSES = {
+    "words": ["TextRoot"],
+    "choice": ["RadioRoot", "SelectRoot"],
+    "checkboxes": ["CheckboxRoot"],
+    "date": ["DateDateInputs"],
+    "time": ["TimeRoot"],
+}
+def question_type(question):
+    for type, classes in QUESTION_CLASSES.items():
+        for class_ in classes:
+            class_ = f"div.freebirdFormviewerComponentsQuestion{class_}"
+            if question.select_one(class_):
+                return type
+    else:
+        raise ValueError("Unknown type of question")
+
+# Return body > script (FB_PUBLIC_LOAD_DATA_)
+def form_json_data(soup):
+    import json
+    script = soup.select("body>script")[0].contents[0]
+    data = script.partition("=")[2].rstrip().removesuffix(";")
+    return json.loads(data)
+
+# Return all input[type=hidden]
+# Not working currently (must run JS, only in browser)
+def form_raw_keys(form):
+    raw_keys = []
+    for input in form.select("input[type=hidden]"):
+        name = input["name"]
+        if "entry" not in name:
+            continue
+        raw_key = name.removeprefix("entry.").partition("_")[0]
+        if raw_key not in raw_keys:
+            raw_keys.append(raw_key)
+    return raw_keys
+
+# Return the inputs' keys ordered
+# Not working currently (must run JS, only in browser)
+def order_keys(raw_keys, types):
+    # Order: other, time, date, radio/checkbox
+    others, times, dates, choices = [], [], [], []
+    for index, type in enumerate(types):
+        if type == "time":
+            times.append(index)
+        elif type == "date":
+            dates.append(index)
+        elif type == "choice":
+            choices.append(index)
+        else:
+            others.append(index)
+    indices = [*others, *times, *dates, *choices]
+    assert len(indices) == len(types)
+    keys = [None] * len(types)
+    for index, raw_key in zip(indices, raw_keys):
+        keys[index] = raw_key
+    return keys
+
+# Return whether the form takes an x-emailAddress
+def form_takes_email(form):
+    selection = (
+        "div.freebirdFormviewerComponentsQuestionBaseRoot "
```

Comment:
Maybe add a constant `FREEBIRD` set to `"freebirdFormviewerComponentsQuestion"`. This can shorten the code to `return bool(form.select_one(f"div.{FREEBIRD}BaseRoot input[type=email]"))`. (From 5 lines to one!)

Reply:
Improved in eb5563866fc6aefd3464efde6ffed330ec1f161d.